apps/rush-lib/src/logic/buildCache/ProjectBuildCache.ts

```diff
@@ -358,80 +366,86 @@ export class ProjectBuildCache {
+        } else {
+          outputFilePaths.push(childRelativePath);
+        }
       }
+    }
 
-      const outputFilePathsForFolder: AsyncIterableIterator<string> = this._getPathsInFolder(
-        terminal,
-        symbolicLinkPathCallback,
-        filteredOutputFolderName,
-        `${projectFolderPath}/${filteredOutputFolderName}`
-      );
+    // Handle declared output folders.
+    for (const outputFolder of this._projectOutputFolderNames) {
+      const diskPath: string = `${posixPrefix}/${outputFolder}`;
+      try {
+        const children: FolderItem[] = await FileSystem.readFolderItemsAsync(diskPath);
+        processChildren(outputFolder, diskPath, children);
+        // The folder exists, record it
+        filteredOutputFolderNames.push(outputFolder);
+      } catch (error) {
+        if (!FileSystem.isNotExistError(error as Error)) {
+          throw error;
+        }
 
-      for await (const outputFilePath of outputFilePathsForFolder) {
-        outputFilePaths.push(outputFilePath);
+        // If the folder does not exist, ignore it.
       }
     }
 
-    if (encounteredEnumerationIssue) {
+    // Walk the tree in parallel
+    await Async.forEachAsync(
+      queue,
+      async ([relativePath, diskPath]: [string, string]) => {
+        const children: FolderItem[] = await FileSystem.readFolderItemsAsync(diskPath);
+        processChildren(relativePath, diskPath, children);
+      },
+      {
+        concurrency: 10
+      }
+    );
+
+    if (hasSymbolicLinks) {
+      // Symbolic links do not round-trip safely.
       return undefined;
     }
 
+    // Ensure stable output path order.
+    outputFilePaths.sort();
+
     return {
-      filteredOutputFolderNames,
-      outputFilePaths
+      outputFilePaths,
+      filteredOutputFolderNames
     };
   }
 
-  private async *_getPathsInFolder(
-    terminal: ITerminal,
-    symbolicLinkPathCallback: (path: string) => void,
-    posixPrefix: string,
-    folderPath: string
-  ): AsyncIterableIterator<string> {
-    const folderEntries: FolderItem[] = await FileSystem.readFolderItemsAsync(folderPath);
-    for (const folderEntry of folderEntries) {
-      const entryPath: string = `${posixPrefix}/${folderEntry.name}`;
-      if (folderEntry.isSymbolicLink()) {
-        symbolicLinkPathCallback(entryPath);
-      } else if (folderEntry.isDirectory()) {
-        yield* this._getPathsInFolder(
-          terminal,
-          symbolicLinkPathCallback,
-          entryPath,
-          `${folderPath}/${folderEntry.name}`
-        );
-      } else {
-        yield entryPath;
-      }
-    }
+  private _getTarLogFilePath(): string {
+    return path.join(this._project.projectRushTempFolder, `${this._cacheId}.log`);
   }
 
-  private _getTarLogFilePath(): string {
-    return path.join(this._project.projectRushTempFolder, 'build-cache-tar.log');
+  private _getTempTarFilePath(): string {
+    return path.join(this._project.projectRushTempFolder, `${this._cacheId}.tgz`);
```

Comment:
What if a user's global cache is on a different volume than the repo? Moving this will potentially take some time.

Can we instead append `-temp_${Date.now()}` or something to the final cache file path?